### PR TITLE
Move meshtally to mesh

### PR DESCRIPTION
**Pull request by zxkjack123**
This PR move the `MeshTally` class from `mcnp.py` to `mesh.py`. The `MeshTally` now is regarded as a independent mesh tally container, but not part of  `Meshtal` of MCNP.

The mesh tally data from OpenMC will be supported (in next several PRs) with the modified `MeshTally`. Mesh parameter and flux data will be read from functions in `mcnp.py` or `openmc_utils.py`, and then put to an instance of `MeshTally`.

### CODE REVIEW

File: pyne/mcnp.py
Comment (gonuke) [inline pyne/mcnp.py:2031]: Please add a blank line after docstring for readability - perhaps also in other functions, even though some of the original functions didn't have this extra line.

**Review by gonuke — CHANGES_REQUESTED** [review]
Thanks @zxkjack123 - this LGTM - just a small formatting request

File: pyne/mcnp.py
Comment (zxkjack123) [reply pyne/mcnp.py:2031]: All the docstrings in `mesh.py` and `mcnp.py` are formatted according to PEP257:https://www.python.org/dev/peps/pep-0257/

Major changs:
- Add blank lines after mult-line docstrings
- Shorten some docstrings into one-line docstrings. (No blank line after one-line docstirngs).

**Review by gonuke — APPROVED** [review]

**Comment by gonuke** [comment]
LGTM, but I wonder how it interacts with some of the current work by @ngranda ?

File: pyne/mcnp.py
Comment (pshriwise) [inline pyne/mcnp.py]:
```suggestion
        """Populate the xsdir object by reading the file."""
```

File: pyne/mcnp.py
Comment (pshriwise) [inline pyne/mcnp.py]: Indentation is a little off here.

File: pyne/mcnp.py
Comment (pshriwise) [inline pyne/mcnp.py]: This looks a little awkward. I'd go with a hanging indent.

File: pyne/mcnp.py
Comment (pshriwise) [inline pyne/mcnp.py]: Same comment on hanging indent here.

File: pyne/mcnp.py
Comment (pshriwise) [inline pyne/mcnp.py:2044]: Should we do some type-checking to make sure we're getting an Iterable of 4 strs here?

File: pyne/mcnp.py
Comment (pshriwise) [inline pyne/mcnp.py:2133]: This method doesn't use `self` and can be made a static method.

File: pyne/mcnp.py
Comment (pshriwise) [inline pyne/mcnp.py]:
```suggestion
        Read meshtally results and relative error data.
```

File: pyne/mcnp.py
Comment (pshriwise) [inline pyne/mcnp.py]: I'd put these shape comments in the line above next to the type.

File: pyne/mcnp.py
Comment (pshriwise) [inline pyne/mcnp.py]:
```suggestion
            for j in range(num_ves):
```

File: pyne/mcnp.py
Comment (pshriwise) [inline pyne/mcnp.py]:
```suggestion
        for i in range(num_e_groups):
```

File: pyne/mcnp.py
Comment (pshriwise) [inline pyne/mcnp.py]: I thought `column_idx` was being stored on the class? But here it's passed in.

File: pyne/mcnp.py
Comment (pshriwise) [inline pyne/mcnp.py]:
```suggestion
            for i in range(num_ves):
```

File: pyne/mcnp.py
Comment (pshriwise) [inline pyne/mcnp.py:2203]: It looks like these lines could cause an error if `num_e_groups > 1`.

`res_tot` and `rel_err_tot` are initialized as empty lists. If we get this point we'll get an `AttributeError` when calling `flatten` I think.

File: pyne/mesh.py
Comment (pshriwise) [inline pyne/mesh.py]:
```suggestion
    """This class stores all information from a single mesh tally that
```

File: pyne/mesh.py
Comment (pshriwise) [inline pyne/mesh.py]: Technically a PyNE Mesh.
```suggestion
    stored as attributes and the "mesh" attribute is a PyNE mesh object with all
```

File: pyne/mesh.py
Comment (pshriwise) [inline pyne/mesh.py]: Overall I'd update this docstring to

```
... "mesh" attribute is a PyNE mesh object tagged with all result and relative error data. This class ....
```

File: pyne/mesh.py
Comment (pshriwise) [inline pyne/mesh.py]:
```suggestion
        True if the tally is modified by a dose response function.
```

File: pyne/mesh.py
Comment (pshriwise) [inline pyne/mesh.py:1539]: I'm wondering if some of these types should really be `tuple`s as we generally don't expect them to change.

File: pyne/mesh.py
Comment (pshriwise) [inline pyne/mesh.py:1542]: Same comment about `tuple` vs. `list` here.

File: pyne/mesh.py
Comment (pshriwise) [inline pyne/mesh.py]:
```suggestion
        total results, and total relative error.
```

File: pyne/mesh.py
Comment (pshriwise) [inline pyne/mesh.py]: This should maybe raise `NotImplemented`? Not :100: on that though.

File: tests/test_r2s.py
Comment (pshriwise) [inline tests/test_r2s.py]:
```suggestion
    os.remove(phtn_src + ".h5")
```

File: tests/test_r2s.py
Comment (pshriwise) [inline tests/test_r2s.py]:
```suggestion
    os.remove(phtn_src + ".h5")
```

**Review by pshriwise — CHANGES_REQUESTED** [review]
This looks like a considerable effort @zxkjack123! It's overall a good design change to make this class more generic IMO. 

A couple of thoughts:

  - I wasn't clear on the status of `column_idx`. Docs say it's an attribute, but it's passed into other methods. Let's make the docs and code consistent.

  - ~~I think it would be useful to turn the `Meshtal.create_meshtally` method into a `MesTtally.from_meshtal` or `Meshtally.from_mcnp` classmethod with the logic for parsing the MCNP meshtal file text as part of MeshTally. Then, in the future, support for an OpenMC Meshtally could be added via a `MeshTally.from_openmc` method.~~ Removed after comment from @gonuke below.

**Comment by gonuke** [comment]
> * I think it would be useful to turn the `Meshtal.create_meshtally` method into a `MesTtally.from_meshtal` or `Meshtally.from_mcnp` classmethod with the logic for parsing the MCNP meshtal file text as part of MeshTally. Then, in the future, support for an OpenMC Meshtally could be added via a `MeshTally.from_openmc` method.

The design here intentionally separates these, on my advice... but we can discuss.  I actually think is a good design for many code-specific “decorations” of common functionality.  All MCNP specific stuff can live in `mcnp.py` but referring to common infrastructure - in this case `MeshTally` infrastructure.  As it is, PyNE has a mixture, where the current `MeshTally` is a good example of common infrastructure living in a code-specific file, and `nucname` being the opposite where all code-specific code is in the common file.

**Comment by pshriwise** [comment]
@gonuke ah, yeah, that's true. Kind of a trade-off there. In general, I tend to favor that style of object generation when possible (too much so, maybe). To me, it boils down to a dependency issue as opposed to the location of code-specific logic. While the logic for MCNP doesn't have any external dependencies, adding a `from_openmc` would insert a dependency on OpenMC in `mesh.py`. Not a desirable situation. 

In the end, I think it should say as-is. I'll update my review comment accordingly.

File: pyne/mcnp.py
Comment (ngranda) [inline pyne/mcnp.py]: Is this an attribute here or an attribute of the MeshTally class?

**Review by ngranda — COMMENTED** [review]
LGTM! Thanks @zxkjack123

**Comment by ngranda** [comment]
> LGTM, but I wonder how it interacts with some of the current work by @ngranda ?

I use the Meshtal class in the same way that the r2s script uses it. I do not think these changes would affect this part of the work. For reading the rnucs mesh, I made my own class that is similar to the Meshtal and MeshTally classes but do not depend on either of them.

File: pyne/mcnp.py
Comment (zxkjack123) [reply pyne/mcnp.py:2044]: The type-checking for `tag_names` are added in https://github.com/pyne/pyne/pull/1199/commits/3d21bf4b7ce522fc31dd1f06d217edecb6bf55e9.

File: pyne/mcnp.py
Comment (zxkjack123) [reply pyne/mcnp.py]: Changed in https://github.com/pyne/pyne/pull/1199/commits/86db650ff475201d4cf61d1e8ea1c04015bdaf98.

File: pyne/mcnp.py
Comment (zxkjack123) [reply pyne/mcnp.py:2133]: I kept the `self` here, for one thing, there are two `read_xxx` functions that uses `self`. For another thing, all there readers operates the same `meshtal` file. I think it's better to keep them linked to the same object (`self`).

File: pyne/mcnp.py
Comment (zxkjack123) [reply pyne/mcnp.py]: Yes, `column_idx` should be stored on the class. Modified in https://github.com/pyne/pyne/pull/1199/commits/c2416ec0260a6d04d7a3eb22d696b2160ec8b5ad.

File: pyne/mcnp.py
Comment (zxkjack123) [reply pyne/mcnp.py:2203]: Well, it is the `result` that calls `flatten()`, but not `res_tot`.
The `result` is initialized by `result = np.empty(shape=(num_e_groups, num_ves))` and then filled with data.

File: pyne/mesh.py
Comment (zxkjack123) [reply pyne/mesh.py]: Updated in https://github.com/pyne/pyne/pull/1199/commits/e50076d1077c74b213305abdacbc9ceb5ff8e393.

File: pyne/mcnp.py
Comment (pshriwise) [reply pyne/mcnp.py:2133]: If we aren’t using `self` in the method to access or set attributes, it doesn’t change how the information is linked to self or how the file handle, `f` is managed. And that method would still maintain it’s context as a reader by nature of being part of the `Meshtal` class definition.

All that being said, `column_idx` is now one of the class attributes (a good update). So, yes, we should keep `self` here. 😃 

We’ll want to update the docstring of this method though as it still indicates that we’re returning a `dict`.

File: pyne/mesh.py
Comment (zxkjack123) [reply pyne/mesh.py:1539]: The `x/y/z_bounds` may involve unit conversion. As far as I know, MCNP and OpenMC use `cm` as length unit. But I am not sure whether there is a MC code use `mm` or `m` as unit. And the mesh bounds transformation could also be used to support a moving component/mesh.

But `e_bounds` involves unit conversion, as MCNP uses `MeV` and OpenMC uses `eV`. So, `e_bounds` should be `list` rather than `tuple`.

File: pyne/mcnp.py
Comment (pshriwise) [reply pyne/mcnp.py:2203]: Ah, so true. I definitely didn’t read that correctly. Thanks!

Can we ensure that the `res_tot` and `res_err_tot` containers are consistent? In one case they will be lists, and in the other they will be NumPy arrays. Unless I’m missing something else!

File: pyne/mesh.py
Comment (zxkjack123) [reply pyne/mesh.py]: Thsi `RuntimeError` is replaced with `NotImplementedError` in https://github.com/pyne/pyne/pull/1199/commits/104e9daead65cf68dd2a9fc8f1bf5175bc8f8b90. This link maybe useful for this type of error: https://help.semmle.com/wiki/display/PYTHON/NotImplemented+is+not+an+Exception.

File: pyne/mcnp.py
Comment (zxkjack123) [reply pyne/mcnp.py]: These two lines are deleted in https://github.com/pyne/pyne/pull/1199/commits/bae3bf1d7e03f507e8729edd3197c8daf20853e5, because it is not an input parameter. The `particle` is an attribute of the `MeshTally` class, but not the `Meshtal`. In theory, a `meshtal` file could contain meshtallies of different particles.

**Comment by zxkjack123** [comment]
> > LGTM, but I wonder how it interacts with some of the current work by @ngranda ?
> 
> I use the Meshtal class in the same way that the r2s script uses it. I do not think these changes would affect this part of the work. For reading the rnucs mesh, I made my own class that is similar to the Meshtal and MeshTally classes but do not depend on either of them.

Glad that this PR does not conflicts with your work.

File: pyne/mcnp.py
Comment (zxkjack123) [reply pyne/mcnp.py:2133]: Docstring of this function is update in https://github.com/pyne/pyne/pull/1199/commits/bae3bf1d7e03f507e8729edd3197c8daf20853e5.

File: pyne/mcnp.py
Comment (zxkjack123) [reply pyne/mcnp.py:2203]: They are convert to `np.array` before return in https://github.com/pyne/pyne/pull/1199/commits/c8c6cdc1ece48782d7ec5a096de2bc3c069c97f0.

File: pyne/mcnp.py
Comment (gonuke) [inline pyne/mcnp.py]: This might be simpler as separate if statements that each raise an exception upon failure:

```
if not check_iteraable(tag_names):
    raise ValueError("....")

if len(tag_names) != 4:
    raise ValueError("....")

````
etc

**Review by gonuke — COMMENTED** [review]
One small suggestion, but not worth blocking the PR over this

**Comment by gonuke** [comment]
If @pshriwise is happy with your changes - we can probably merge

File: pyne/mcnp.py
Comment (zxkjack123) [reply pyne/mcnp.py]: Thanks. Suggestion adopted in https://github.com/pyne/pyne/pull/1199/commits/f4b6dccff3290e3d0bb6e6a646e029199c95c759.

**Review by pshriwise — APPROVED** [review]
Looks ready to me. As usual, thanks for the diligence @zxkjack123!

File: pyne/mcnp.py
Comment (gonuke) [inline pyne/mcnp.py]:
```suggestion
    # tag_names should be a string with length of 4
```

**Comment by gonuke** [comment]
Not sure why the tests are failing!??!

**Comment by pshriwise** [comment]
Something weird with the config file?

This run has a lot of it commented: https://circleci.com/gh/zxkjack123/pyne/5152#config/containers/0

But the run from the commit before looks fine:
https://circleci.com/gh/zxkjack123/pyne/5151#config/containers/0

Not sure what the root cause is though. Not a circleci guru. @bam241?

**Comment by gonuke** [comment]
Must have been a failure on the CircleCI end - I'll merge once the tests complete...

**Comment by gonuke** [comment]
Thanks for all your effort here @zxkjack123

**Pull request merged**
